File: javapatterns/src/main/java/org/iut/nantes/state/avion/Etat.java
package org.iut.nantes.state.avion;

import java.util.Date;

abstract class Etat
{
  protected Billet billet;

  public Etat( Billet b )
  {
    billet = b;
  }

  // Transitions
  public void reserver( Date date ) {}
  public void confirmerEtPayer() {}
  public void annuler() {}
  public void decollage() {}
  
  // Action
  public void modifierDate( Date date )
  {
    // action par défaut
    System.out.println("Il est impossible de modifier la date.");
  }
}

File: src/fr/tblf/patterns/state/avion/etats/Etat.kt
package fr.tblf.patterns.state.avion.etats

import fr.tblf.patterns.state.avion.Billet
import java.util.*

abstract class Etat(val billet: Billet) {
    open fun reserver() {}
    open fun annuler() {}
    open fun payer() {}
    open fun partir() {}
    open fun modifier(date: Date) {}
}

class Disponible(billet: Billet) : Etat(billet) {
    override fun reserver() {
        println("Reservation effectuée!")
        billet.changeEtatReserve()
    }
}

class Reserve(billet: Billet) : Etat(billet) {
    override fun annuler() {
        println("Billet annulé !")
        billet.changeEtatDisponible()
    }

    override fun payer() {
        println("Paiement effectué ! Le billet est maintenant commandé.")
        billet.changeEtatConfirme()
    }

    override fun modifier(date: Date) {
        println("Date mise à jour au: $date")
        billet.changerDate(date)
    }
}

class Confirme(billet: Billet) : Etat(billet) {
    override fun modifier(date: Date) {
        println("Date mise à jour : $date")
        billet.changerDate(date)
    }

    override fun partir() {
        println("Décollage !")
        billet.changeEtatConsomme()
    }
}

class Consomme(billet: Billet) : Etat(billet) {

}